src/parser/mod.rs

```diff
@@ -4535,7 +4535,14 @@ impl<'a> Parser<'a> {
         self.expect_token(&Token::RParen)?;
 
         let return_type = if self.parse_keyword(Keyword::RETURNS) {
-            Some(self.parse_data_type()?)
+            if dialect_of!(self is PostgreSqlDialect | GenericDialect)
+                && self.parse_keyword(Keyword::TABLE)
+            {
+                let columns = self.parse_parenthesized_columns()?;
+                Some(DataType::Table(columns))
+            } else {
+                Some(self.parse_data_type()?)
```

Comment:
I'm thinking the behavior here could be a bit misleading, given the variant `DataType::Table`, the expectation is that `self.parse_data_type()` can parse that variant. Would that be reasonable to do here instead (i.e. the table type logic is moved into `parse_data_type()` and this part becomes unchanged [likely no need to guard by dialect either since the syntax doesn't conflict] I imagine as a result)?

Reply:
Yes sure, I was not sure that adding it to the parse_data_type() would not have side effects on other parts of the parser as the function is generic and not focused only on the function case.